share/wake/lib/core/examples.wake

```diff
@@ -0,0 +1,191 @@
+# Copyright 2021 SiFive, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You should have received a copy of LICENSE.Apache2 along with
+# this software. If not, you may obtain a copy at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+package wake
+
+def expectOne (msg: String) (result: Result (List a) Error): Result a Error = match result
+    Fail err      = addErrorContext msg (Fail err)
+    Pass (x, Nil)  = Pass x
+    Pass otherwise = failWithError "{msg}: expected 1, got {len otherwise | str}"
+
+# Ported from the jq tutorial
+export def example1 (j: JValue): Result JValue Error = Pass j
+export def example2 (j: JValue): Result JValue Error =
+    j | jAt 0 | expectOne ""
+
+# Simplify github json first commit
+export def example3 (log: JValue): Result JValue Error =
+    require Pass commit =
+        single log
+        | Pass
+        |> jAt 0
+        | expectOne "Head of log not found"
+
+    require Pass message =
+        single commit
+        | Pass
+        |> jField "commit"
+        |> jField "message"
+        | expectOne "commit message not found"
+
+    require Pass name =
+        single commit
+        | Pass
+        |> jField "commit"
+        |> jField "name"
+        | expectOne "commit author not found"
+
+    def output = JObject (
+        "message" :-> message,
+        "name" :-> name,
+        Nil
+    )
+
+    Pass output
+
+# Simplify full github json log
+export def example4 (log: JValue): Result JValue Error =
+    def simpleCommit commit =
+        require Pass message =
+            single commit
+            | Pass
+            |> jField "commit"
+            |> jField "message"
+            | expectOne ".commit.message missing"
+
+        require Pass name =
+            single commit
+            | Pass
+            |> jField "commit"
+            |> jField "name"
+            | expectOne ".commit.name missing"
+
+        def output = JObject (
+            "message" :-> message,
+            "name" :-> name,
+            Nil
+        )
+
+        Pass output
+
+    single log
+    | Pass
+    |> jFlatten
+    |: findFailFn simpleCommit
+    | rmap JArray
```

Comment:
Or, just:
```
require Pass orig = jq log | jFlatten
require Pass simplified = findFailFn simpleCommit orig
Pass (JArray simplified)
```

Reply:
Alternatively, we could defined `jMap` with type `(JValue => Result JValue Error) => Result (List JValue) Error`:

```
require Pass simplified =
  jq log
  | jFlatten
  | jMap simpleCommit

Pass (JArray simplified)
```